File: macos/studio/Sources/DreamSkinStudioCore/EngineProtocol.swift
import Foundation

public enum EngineOperation: String, Codable, CaseIterable, Sendable {
    case preflight
    case install
    case apply
    case status
    case pause
    case resume
    case restore
    case verify
    case uninstall
}

public enum EngineProgress: String, Codable, CaseIterable, Sendable {
    case checking
    case preparing
    case installing
    case launching
    case connecting
    case applying
    case verifying
    case pausing
    case restoring
    case uninstalling
}

public struct EngineState: Codable, Equatable, Sendable {
    public enum Install: String, Codable, CaseIterable, Sendable {
        case notInstalled = "not-installed"
        case ready
    }

    public enum Codex: String, Codable, CaseIterable, Sendable {
        case notInstalled = "not-installed"
        case needsFirstRun = "needs-first-run"
        case stopped
        case running
    }

    public enum Session: String, Codable, CaseIterable, Sendable {
        case official
        case active
        case paused
        case stale
    }

    public enum Operation: String, Codable, CaseIterable, Sendable {
        case idle
        case busy
    }

    public enum Action: String, Codable, CaseIterable, Sendable {
        case install
        case apply
        case pause
        case resume
        case restore
        case verify
        case uninstall
    }

    public let install: Install
    public let codex: Codex
    public let session: Session
    public let operation: Operation
    public let themeName: String?
    public let requiresRestart: Bool
    public let availableActions: [Action]
    public let verified: Bool?

    private enum CodingKeys: String, CodingKey, CaseIterable {
        case install
        case codex
        case session
        case operation
        case themeName
        case requiresRestart
        case availableActions
        case verified
    }

    public init(from decoder: any Decoder) throws {
        let container = try decoder.strictContainer(keys: CodingKeys.allCases.map(\.rawValue))
        install = try container.decode(Install.self, forKey: "install")
        codex = try container.decode(Codex.self, forKey: "codex")
        session = try container.decode(Session.self, forKey: "session")
        operation = try container.decode(Operation.self, forKey: "operation")
        themeName = try container.decodeRequiredNullable(String.self, forKey: "themeName")
        requiresRestart = try container.decode(Bool.self, forKey: "requiresRestart")
        availableActions = try container.decode([Action].self, forKey: "availableActions")
        verified = try container.decodeRequiredNullable(Bool.self, forKey: "verified")
        try requireUnique(availableActions, decoder: decoder)
        guard session != .active || (install == .ready && codex == .running) else {
            throw protocolCorrupted(decoder)
        }
    }
}

public struct EngineError: Codable, Equatable, Sendable {
    public enum Code: String, Codable, CaseIterable, Sendable {
        case invalidRequest = "INVALID_REQUEST"
        case operationBusy = "OPERATION_BUSY"
        case codexNotInstalled = "CODEX_NOT_INSTALLED"
        case codexFirstRunRequired = "CODEX_FIRST_RUN_REQUIRED"
        case codexIdentityInvalid = "CODEX_IDENTITY_INVALID"
        case runtimeInvalid = "RUNTIME_INVALID"
        case codexCloseRequired = "CODEX_CLOSE_REQUIRED"
        case restartRequired = "RESTART_REQUIRED"
        case forceStopRequired = "FORCE_STOP_REQUIRED"
        case stateUnsafe = "STATE_UNSAFE"
        case portUnavailable = "PORT_UNAVAILABLE"
        case configUnsafe = "CONFIG_UNSAFE"
        case configChanged = "CONFIG_CHANGED"
        case configBackupMissing = "CONFIG_BACKUP_MISSING"
        case themeInvalid = "THEME_INVALID"
        case injectorFailed = "INJECTOR_FAILED"
        case verifyFailed = "VERIFY_FAILED"
        case liveRemoveFailed = "LIVE_REMOVE_FAILED"
        case operationFailed = "OPERATION_FAILED"
        case internalError = "INTERNAL_ERROR"
    }

    public enum RecoveryAction: String, Codable, CaseIterable, Sendable {
        case openCodex = "open-codex"
        case authorizeRestart = "authorize-restart"
        case authorizeForceStop = "authorize-force-stop"
        case retry
        case restore
        case diagnostics
        case cancel
    }

    public let code: Code
    public let message: String
    public let recoveryActions: [RecoveryAction]

    private enum CodingKeys: String, CodingKey, CaseIterable {
        case code
        case message
        case recoveryActions
    }

    public init(from decoder: any Decoder) throws {
        let container = try decoder.strictContainer(keys: CodingKeys.allCases.map(\.rawValue))
        code = try container.decode(Code.self, forKey: "code")
        message = try container.decode(String.self, forKey: "message")
        recoveryActions = try container.decode([RecoveryAction].self, forKey: "recoveryActions")
        guard !message.isEmpty else { throw protocolCorrupted(decoder) }
        try requireUnique(recoveryActions, decoder: decoder)
    }
}

public struct EngineEnvelope: Codable, Equatable, Sendable {
    public let schemaVersion: Int
    public let ok: Bool
    public let operation: EngineOperation
    public let state: EngineState
    public let error: EngineError?

    private enum CodingKeys: String, CodingKey, CaseIterable {
        case schemaVersion
        case ok
        case operation
        case state
        case error
    }

    public init(from decoder: any Decoder) throws {
        let container = try decoder.strictContainer(keys: CodingKeys.allCases.map(\.rawValue))
        schemaVersion = try container.decode(Int.self, forKey: "schemaVersion")
        ok = try container.decode(Bool.self, forKey: "ok")
        operation = try container.decode(EngineOperation.self, forKey: "operation")
        state = try container.decode(EngineState.self, forKey: "state")
        error = try container.decodeRequiredNullable(EngineError.self, forKey: "error")

        guard schemaVersion == 1, ok == (error == nil) else {
            throw protocolCorrupted(decoder)
        }
        if ok {
            switch operation {
            case .apply, .resume, .verify:
                guard state.verified == true else { throw protocolCorrupted(decoder) }
            default:
                break
            }
        }
    }
}

public protocol EngineRunning: Sendable {
    func run(
        _ operation: EngineOperation,
        restartAuthorized: Bool,
        forceAuthorized: Bool,
        deleteUserThemes: Bool,
        onProgress: @escaping @Sendable (EngineProgress) -> Void
    ) async throws -> EngineEnvelope
}

private struct AnyCodingKey: CodingKey, Hashable {
    let stringValue: String
    let intValue: Int?

    init(_ stringValue: String) {
        self.stringValue = stringValue
        intValue = nil
    }

    init?(stringValue: String) {
        self.init(stringValue)
    }

    init?(intValue: Int) {
        stringValue = String(intValue)
        self.intValue = intValue
    }
}

private extension Decoder {
    func strictContainer(keys: [String]) throws -> KeyedDecodingContainer<AnyCodingKey> {
        let container = try container(keyedBy: AnyCodingKey.self)
        guard Set(container.allKeys.map(\.stringValue)) == Set(keys) else {
            throw protocolCorrupted(self)
        }
        return container
    }
}

private extension KeyedDecodingContainer where Key == AnyCodingKey {
    func decode<Value: Decodable>(_ type: Value.Type, forKey key: String) throws -> Value {
        try decode(type, forKey: AnyCodingKey(key))
    }

    func decodeRequiredNullable<Value: Decodable>(_ type: Value.Type, forKey key: String) throws -> Value? {
        let codingKey = AnyCodingKey(key)
        guard contains(codingKey) else {
            throw DecodingError.keyNotFound(
                codingKey,
                DecodingError.Context(codingPath: codingPath, debugDescription: "Missing protocol key.")
            )
        }
        return try decodeIfPresent(type, forKey: codingKey)
    }
}

private func requireUnique<Value: Hashable>(_ values: [Value], decoder: any Decoder) throws {
    guard Set(values).count == values.count else { throw protocolCorrupted(decoder) }
}

private func protocolCorrupted(_ decoder: any Decoder) -> DecodingError {
    .dataCorrupted(DecodingError.Context(
        codingPath: decoder.codingPath,
        debugDescription: "Invalid Studio protocol envelope."
    ))
}

File: windows/studio/EngineProtocol.cs
using System.IO;
using System.Text.Json;

namespace CodexDreamSkinStudio;

internal enum EngineOperation { Preflight, Install, Apply, Status, Pause, Resume, Restore, Verify, Uninstall }
internal enum EngineProgress { Checking, Preparing, Installing, Launching, Connecting, Applying, Verifying, Pausing, Restoring, Uninstalling }
internal sealed record EngineState(string Install, string Codex, string Session, string Operation, string? ThemeName, bool RequiresRestart, string[] AvailableActions, bool? Verified);
internal sealed record EngineError(string Code, string Message, string[] RecoveryActions);
internal sealed record EngineEnvelope(int SchemaVersion, bool Ok, string Operation, EngineState State, EngineError? Error);

internal static class EngineOperationExtensions
{
  internal static string ToArgument(this EngineOperation operation) => operation.ToString().ToLowerInvariant();
}

internal static class EngineProtocol
{
  internal static readonly JsonSerializerOptions JsonOptions = new() { PropertyNameCaseInsensitive = true };
  private static readonly HashSet<string> Operations = Enum.GetNames<EngineOperation>().Select(x => x.ToLowerInvariant()).ToHashSet();
  private static readonly HashSet<string> ProgressValues = Enum.GetNames<EngineProgress>().Select(x => x.ToLowerInvariant()).ToHashSet();
  private static readonly HashSet<string> Installs = ["not-installed", "ready"];
  private static readonly HashSet<string> CodexStates = ["not-installed", "needs-first-run", "stopped", "running"];
  private static readonly HashSet<string> Sessions = ["official", "active", "paused", "stale"];
  private static readonly HashSet<string> StateOperations = ["idle", "busy"];
  private static readonly HashSet<string> Actions = ["install", "apply", "pause", "resume", "restore", "verify", "uninstall"];
  private static readonly HashSet<string> RecoveryActions = ["open-codex", "retry", "cancel", "restore", "diagnostics", "authorize-restart", "authorize-force-stop"];
  private static readonly HashSet<string> ErrorCodes = [
    "INVALID_REQUEST", "OPERATION_BUSY", "CODEX_NOT_INSTALLED", "CODEX_FIRST_RUN_REQUIRED", "CODEX_IDENTITY_INVALID",
    "RUNTIME_INVALID", "CODEX_CLOSE_REQUIRED", "RESTART_REQUIRED", "FORCE_STOP_REQUIRED", "STATE_UNSAFE",
    "PORT_UNAVAILABLE", "CONFIG_UNSAFE", "CONFIG_CHANGED", "CONFIG_BACKUP_MISSING", "THEME_INVALID",
    "INJECTOR_FAILED", "VERIFY_FAILED", "LIVE_REMOVE_FAILED", "OPERATION_FAILED", "INTERNAL_ERROR"
  ];

  internal static EngineEnvelope Parse(string stdout, EngineOperation requestedOperation)
  {
    if (string.IsNullOrEmpty(stdout) || stdout[0] == '\uFEFF') throw new InvalidDataException("The engine response is not valid Protocol v1 output.");

    JsonDocument document;
    try
    {
      document = JsonDocument.Parse(stdout, new JsonDocumentOptions { CommentHandling = JsonCommentHandling.Disallow, AllowTrailingCommas = false });
    }
    catch (JsonException) { throw; }

    using (document)
    {
      var root = document.RootElement;
      RequireObject(root, ["schemaVersion", "ok", "operation", "state", "error"]);
      RequireKind(root, "schemaVersion", JsonValueKind.Number);
      RequireKind(root, "ok", JsonValueKind.True, JsonValueKind.False);
      RequireKind(root, "operation", JsonValueKind.String);
      RequireKind(root, "state", JsonValueKind.Object);
      if (Get(root, "error").ValueKind is not (JsonValueKind.Null or JsonValueKind.Object)) Invalid();

      var state = Get(root, "state");
      RequireObject(state, ["install", "codex", "session", "operation", "themeName", "requiresRestart", "availableActions", "verified"]);
      RequireString(state, "install");
      RequireString(state, "codex");
      RequireString(state, "session");
      RequireString(state, "operation");
      if (Get(state, "themeName").ValueKind is not (JsonValueKind.Null or JsonValueKind.String)) Invalid();
      RequireKind(state, "requiresRestart", JsonValueKind.True, JsonValueKind.False);
      RequireKind(state, "availableActions", JsonValueKind.Array);
      if (Get(state, "verified").ValueKind is not (JsonValueKind.Null or JsonValueKind.True or JsonValueKind.False)) Invalid();

      var error = Get(root, "error");
      if (error.ValueKind == JsonValueKind.Object)
      {
        RequireObject(error, ["code", "message", "recoveryActions"]);
        RequireString(error, "code");
        RequireString(error, "message");
        RequireKind(error, "recoveryActions", JsonValueKind.Array);
        ValidateStringArray(Get(error, "recoveryActions"), RecoveryActions);
        if (Get(error, "code").GetString() is not { } errorCode || !ErrorCodes.Contains(errorCode) ||
            string.IsNullOrWhiteSpace(Get(error, "message").GetString())) Invalid();
      }

      EngineEnvelope envelope;
      try { envelope = JsonSerializer.Deserialize<EngineEnvelope>(root.GetRawText(), JsonOptions) ?? throw new InvalidDataException(); }
      catch (JsonException exception) { throw new InvalidDataException("The engine response is incomplete.", exception); }

      if (envelope.SchemaVersion != 1 ||
          !Operations.Contains(envelope.Operation) ||
          envelope.Operation != requestedOperation.ToArgument() ||
          !Installs.Contains(envelope.State.Install) ||
          !CodexStates.Contains(envelope.State.Codex) ||
          !Sessions.Contains(envelope.State.Session) ||
          !StateOperations.Contains(envelope.State.Operation) ||
          envelope.Ok != (envelope.Error is null)) Invalid();
      ValidateStringArray(Get(state, "availableActions"), Actions);
      if (envelope.State.Operation == "busy" && envelope.State.AvailableActions.Length != 0) Invalid();
      if (envelope.State.Session == "active" && (envelope.State.Install != "ready" || envelope.State.Codex != "running")) Invalid();
      if (envelope.State.Verified == true && envelope.State.Session != "active") Invalid();
      if (envelope.Ok && requestedOperation is EngineOperation.Apply or EngineOperation.Resume or EngineOperation.Verify && envelope.State.Verified != true) Invalid();
      ValidateSemantics(envelope, requestedOperation);
      return envelope;
    }
  }

  internal static void ParseProgress(string stderr, IProgress<EngineProgress>? progress)
  {
    if (stderr.Length == 0) return;
    var normalized = stderr.Replace("\r\n", "\n");
    if (normalized.EndsWith('\n')) normalized = normalized[..^1];
    foreach (var line in normalized.Split('\n'))
    {
      const string prefix = "DREAM_SKIN_PROGRESS ";
      if (line.Length == 0 || !line.StartsWith(prefix, StringComparison.Ordinal) || !ProgressValues.Contains(line[prefix.Length..])) Invalid();
      progress?.Report(Enum.Parse<EngineProgress>(line[prefix.Length..], true));
    }
  }

  private static void RequireObject(JsonElement value, string[] expected)
  {
    if (value.ValueKind != JsonValueKind.Object) Invalid();
    var properties = value.EnumerateObject().ToArray();
    if (properties.Length != expected.Length || properties.Select(x => x.Name).Distinct(StringComparer.OrdinalIgnoreCase).Count() != properties.Length ||
        properties.Any(x => !expected.Contains(x.Name, StringComparer.OrdinalIgnoreCase))) Invalid();
  }

  private static JsonElement Get(JsonElement value, string name)
  {
    foreach (var property in value.EnumerateObject())
      if (property.Name.Equals(name, StringComparison.OrdinalIgnoreCase)) return property.Value;
    Invalid();
    return default;
  }

  private static void RequireKind(JsonElement value, string name, params JsonValueKind[] kinds)
  {
    if (!kinds.Contains(Get(value, name).ValueKind)) Invalid();
  }

  private static void RequireString(JsonElement value, string name)
  {
    RequireKind(value, name, JsonValueKind.String);
    if (Get(value, name).GetString() is not { Length: > 0 }) Invalid();
  }

  private static void ValidateStringArray(JsonElement values, HashSet<string> allowed)
  {
    var seen = new HashSet<string>(StringComparer.Ordinal);
    foreach (var value in values.EnumerateArray())
      if (value.ValueKind != JsonValueKind.String || value.GetString() is not { } text || !allowed.Contains(text) || !seen.Add(text)) Invalid();
  }

  private static void ValidateSemantics(EngineEnvelope envelope, EngineOperation requestedOperation)
  {
    var state = envelope.State;
    var actions = state.AvailableActions;
    var safeUnavailableStatusFailure = !envelope.Ok && requestedOperation is EngineOperation.Preflight or EngineOperation.Status &&
      envelope.Error?.Code == "INTERNAL_ERROR" && envelope.Error.RecoveryActions.SequenceEqual(["retry", "diagnostics", "cancel"]) &&
      state.Install == "not-installed" && state.Codex == "not-installed" && state.Session == "stale" &&
      state.Operation == "idle" && state.ThemeName is null && !state.RequiresRestart && actions.Length == 0 && state.Verified is null;
    var safeUnavailableStatus = requestedOperation is EngineOperation.Preflight or EngineOperation.Status &&
      state.Install == "not-installed" && state.Session == "official" && state.Operation == "idle" &&
      actions.SequenceEqual(["install"]) && state.Verified is null;
    if (state.RequiresRestart && (state.Codex != "running" || state.Session != "official")) Invalid();
    if (state.Session != "official" && state.Install != "ready" && !safeUnavailableStatusFailure) Invalid();
    if (state.Operation == "busy")
    {
      if (envelope.Ok || envelope.Error?.Code != "OPERATION_BUSY" || actions.Length != 0) Invalid();
    }
    else if (envelope.Error?.Code == "OPERATION_BUSY") Invalid();
    if (envelope.Error?.Code == "INVALID_REQUEST" && (state.Install != "not-installed" || state.Codex != "not-installed" ||
        state.Session != "official" || state.Operation != "idle" || state.ThemeName is not null || state.RequiresRestart ||
        state.Verified is not null || actions.Length != 0)) Invalid();

    if (state.Install == "ready")
    {
      if (actions.Contains("install")) Invalid();
    }
    else
    {
      if (actions.Any(action => action != "install") || state.Verified is not null ||
          state.ThemeName is not null && !safeUnavailableStatus) Invalid();
    }
    if (actions.Contains("pause") && state.Session != "active") Invalid();
    if (actions.Contains("resume") && state.Session is not ("active" or "paused")) Invalid();
    if (actions.Contains("apply") && state.Session == "active") Invalid();

    if (!envelope.Ok) return;
    if (state.Install == "not-installed" && requestedOperation is not EngineOperation.Uninstall && !safeUnavailableStatus) Invalid();
    if (state.Operation != "idle") Invalid();
    switch (requestedOperation)
    {
      case EngineOperation.Install when state.Install != "ready":
      case EngineOperation.Pause when state.Session != "paused":
      case EngineOperation.Restore when state.Session != "official":
        Invalid();
        break;
      case EngineOperation.Uninstall when state.Install != "not-installed" || state.Codex != "stopped" ||
        state.Session != "official" || state.ThemeName is not null || state.Verified is not null ||
        !actions.SequenceEqual(["install"]):
        Invalid();
        break;
    }
  }

  private static void Invalid() => throw new InvalidDataException("The engine response violates Protocol v1.");
}
